code/datums/components/storage.dm

```diff
@@ -0,0 +1,335 @@
+					ok = 1
+		if (!ok)
+			boutput(user, "<span class='alert'>[source] cannot hold [I].</span>")
+			return
+
+	else if (I.w_class > src.max_wclass)
+		boutput(user, "<span class='alert'>[I] won't fit into [source]!</span>")
+		return
+
+	if (get_contents_len(source) >= slots)
+		boutput(usr, "<span class='alert'>[source] is full!</span>")
+		return
+	var/atom/checkloc = source.loc
+	while (checkloc && !isturf(source.loc))
+		if (checkloc == I)
+			return
+		checkloc = checkloc.loc
+
+	src.add_item(source, I, user)
+
+	source.add_fingerprint(user)
+	animate_storage_rustle(source)
+	if (!src.sneaky && !istype(I, /obj/item/gun/energy/crossbow))
+		user.visible_message("<span class='notice'>[user] has added [I] to [source]!</span>", "<span class='notice'>You have added [I] to [source].</span>")
+	playsound(source.loc, "rustle", 50, 1, -5)
+	return COMSIG_RETURN_EARLY
+
+// source attacked by user with an empty hand
+/datum/component/storage/proc/hand_attack(source, mob/user)
+	return src.handle_storage(user)
+
+// source click dragged to over_object (can be HUD or usr)
+/datum/component/storage/proc/mouse_drop(atom/source, atom/over_object)
+	var/obj/screen/hud/S = over_object
+	if (istype(S))
+		playsound(source.loc, "rustle", 50, 1, -5)
+		if (!usr.restrained() && !usr.stat && source.loc == usr)
+			if (S.id == "rhand")
+				if (!usr.r_hand)
+					usr.u_equip(source)
+					usr.put_in_hand(source, 0)
+			else
+				if (S.id == "lhand")
+					if (!usr.l_hand)
+						usr.u_equip(source)
+						usr.put_in_hand(source, 1)
+			return
+	if (over_object == usr && in_range(source, usr) && isliving(usr) && !usr.stat)
+		if (usr.s_active)
+			usr.detach_hud(usr.s_active)
+			usr.s_active = null
+		if (src.mousetrap_check(usr))
+			return
+		usr.s_active = hud
+		hud.update()
+		usr.attach_hud(hud)
+		return
+	if (usr.is_in_hands(source))
+		var/turf/T = over_object
+		if (istype(T, /obj/table))
+			T = get_turf(T)
+		if (!(usr in range(1, T)))
+			return
+		if (istype(T))
+			for (var/obj/O in T)
+				if (O.density && !istype(O, /obj/table) && !istype(O, /obj/rack))
+					return
+			if (!T.density)
+				usr.visible_message("<span class='alert'>[usr] dumps the contents of [source] onto [T]!</span>")
+				for (var/obj/item/I in source.contents)
+					I.set_loc(T)
+					I.layer = initial(I.layer)
+					if (istype(I, /obj/item/mousetrap))
+						var/obj/item/mousetrap/MT = I
+						if (MT.armed)
+							MT.visible_message("<span class='alert'>[MT] triggers as it falls on the ground!</span>")
+							MT.triggered(usr, null)
+					else if (istype(I, /obj/item/mine))
+						var/obj/item/mine/M = I
+						if (M.armed && M.used_up != 1)
+							M.visible_message("<span class='alert'>[M] triggers as it falls on the ground!</span>")
+							M.triggered(usr)
+					hud.remove_item(I)
+
+/datum/component/storage/proc/move_trigger(atom/source, mob/user, kindof)
+	for (var/obj/O in source.contents)
+		if (O.move_triggered)
+			O.move_trigger(user, kindof)
```

Comment:
Out of the scope of this PR but this should go at some point. Or rather objects that do have the `move_triggered` flag turned on (at the moment that's literally just reagent containers containing nitroglycerin) should pass that up the contents chain because currently whenever you make a step every single container you have on your person gets recursively notified.